Communicate READER to merlin for configured dialects

andreypopp opened the pull request:
Communicate `READER` to merlin for configured dialects.

The user visible change is the `(merlin_reader ...)` is now allowed in `dune-project` in `(dialect ...)` configuration:
```
(dialect
 (name mlx)
 (implementation
  (extension mlx)
  (merlin_reader mlx)
  (preprocess
   (run ./mlx/pp.exe %{input-file}))))
```

rgrinberg commented:
cc @nojb who is the main author of the dialects feature.

Could you add some tests please? See our merlin tests for how we test the communication with merlin.

nojb commented:
Thanks for the ping @rgrinberg. @andreypopp: I am not familiar with the READER feature. Do you mind explaining what it is about? Thanks!

nojb commented:
> @nojb in merlin there's a way to plug a custom "reader" which allows to customise parsing (a few other features). The usually way to implement reader is to use https://github.com/let-def/merlin-extend SDK.

Thanks. Do I understand correctly that a reader is an executable? In any case, it seems like a coherent addition to the `(dialect)` stanza.

nojb commented on src/dune_rules/dialect.ml:
```diff
@@ -78,13 +81,14 @@ let decode =
       field_o
         "format"
         (map ~f:(fun (loc, x) -> loc, x, []) (located Action.decode_dune_file))
+    and+ merlin_reader = field_o "merlin_reader" (located (repeat1 string))
```
This field should be versioned.

andreypopp commented:
> Thanks. Do I understand correctly that a reader is an executable?

The reader itself is an executable, yes, the configuration though specifies just a reader's NAME, then merlin seaches for `ocamlmerlin-NAME` in `$PATH`.

nojb commented on src/dune_rules/dialect.ml:
```diff
@@ -78,13 +81,24 @@ let decode =
+        let what = "the possibility of defining extensions containing periods" in
+        Syntax.Error.since loc Stanza.syntax ver ~what)
+    in
+    let () =
+      let ver = 3, 10 in
+      match syntax_ver < ver, merlin_reader with
+      | true, Some (loc, _) ->
+        let what = "the (merlin_reader) stanza" in
+        Syntax.Error.since loc Stanza.syntax ver ~what
+      | _ -> ()
+    in
```
The error handling check that you imitated here is specially tricky; in your case you can do something simpler, such as in https://github.com/ocaml/dune/blob/0219fad9ae3d7e45e5cd37616bfb126b9c96d802/src/dune_rules/dune_file.ml#L76-L80

andreypopp commented:
Actually I'm wondering if we should allow to define a reader per `dialect` or per `dialect * ml_kind`?

I've made it per `dialect * ml_kind` as merlin maps extensions to readers but, in practice I think, both implementations and interfaces will be handled by a single executable so maybe we should move this stanza up to be per `dialect`? What do you think?

nojb commented on src/dune_rules/dialect.ml:
```diff
@@ -227,21 +251,26 @@ module DB = struct
+
+  and for_merlin =
+    { extensions : string option Ml_kind.Dict.t list
+    ; extension_to_reader : (string * string list) list
```
If I understand correctly, this is supposed to map file extensions to lists of "readers". Wouldn't it be better to use a Map here? Also, I find the name a bit unintuitive; what about simply `readers`?

andreypopp replied:
Can use map but in practice the number of dialects is negligible I think...

nojb replied:
I wasn't suggesting it because of the number of dialects, rather because it more clearly exposes the "map" semantics of the field (eg the keys are unique).

andreypopp replied:
I see, that's what the field name was supposedly to clarify. Anyway, I've changed it as suggested (and the version check as well).

nojb commented:
> What do you think?

I think the current approach (`dialect * ml_kind`) gives a bit more flexibility and stays closer to the interface exposed by Merlin, so I would leave it as it is.

nojb commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -51,17 +51,22 @@ module Processed = struct
     ; src_dirs : Path.Set.t
     ; flags : string list
     ; extensions : string option Ml_kind.Dict.t list
+    ; extension_to_reader : string list String.Map.t
```
Shall we call this field `readers` as well?

andreypopp replied:
Yes, sorry, just pushed a commit fixing this.

nojb commented on src/dune_rules/dialect.ml:
```diff
@@ -253,17 +275,24 @@ module DB = struct
+          match merlin_reader d kind with
+          | None -> s
+          | Some (extension, reader) ->
+            (* Ok to use [add_exn] here as we've validating below that we have
```
Do you mean `above` here instead of `below`?

andreypopp replied:
The code which validates extension is registered once is [below](https://github.com/ocaml/dune/pull/8567/files/9669feffd1a3b66e27a7783e5a2c6ed41cd61f8b#diff-e77f32927aca7df771b4e00ecc3315a416133fdc68e1c9133139a60ef399fec1R310). I'll add a more specific reference though, good catch.

nojb reviewed (approved):
I didn't looks in full detail, but generally speaking this LGTM

nojb commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -394,7 +439,7 @@ module Unprocessed = struct
     in
     let flags = Ocaml_flags.get flags mode in
-    let extensions = Dialect.DB.extensions_for_merlin dialects in
+    let for_merlin = Dialect.DB.for_merlin dialects in
```
Nit: could use a record pattern here `let {Dialect.DB.extensions; for_merlin} = ...` and use punning below.

andreypopp commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -257,25 +286,33 @@ module Processed = struct
-        ++ Pp.newline
-        ++ Pp.vbox (Sexp.pp sexp)
+        let sexps =
+          List.map (Module.sources module_) ~f:(fun path ->
```
So this is working incorrectly now as it gets `.ml` suffixed versions, e.g for `App.mlx` it will get `App.mlx.ml` and `App.mlx.mli`.

1. I'm not sure how to fix that... seems like after `Module.ml_source` we don't track the original modules so no info about the dialect or original file name is present here.
2. As I understand this command is purely for tests/debug, maybe we should make it iterate over source files and "modules"?

andreypopp replied:
Just to clarify, in 2. above I'm suggesting to remove `print_file` completely and make `dune ocaml merlin dump-config` work by listing the directory and then querying a config for each source file found.

nojb replied:
I'm not familiar with this part of the code, @rgrinberg or @voodoos may know.

voodoos replied:
I think this would simplified by moving the reader info the the part of the config file that is module-specific instead of the part that is common for all modules of a stanza.

See my comment here: https://github.com/ocaml/dune/pull/8567/files#r1314880623

andreypopp commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -51,17 +51,21 @@ module Processed = struct
     ; src_dirs : Path.Set.t
     ; flags : string list
     ; extensions : string option Ml_kind.Dict.t list
+    ; readers : string list String.Map.t
```
I see, this indeed simplifies things but it doesn't solve the issue with `dune ocaml merlin dump-config` — the place where we create per module configs already sees preprocessed modules only.

andreypopp replied:
So in [4080abe](https://github.com/ocaml/dune/pull/8567/commits/4080abe502d60d750a4d2db482c5d36ca5463890) I've refactored it to pass source modules when configuring merlin.

This seems to me the right thing to do conceptually — as merlin works on source files and thus it makes sense to produce configuration for what's in source/user written modules.


But not sure if I broke some edge cases? From my smoke testing it works fine both for files with dialects and for regular .ml files.

andreypopp commented:
Made the tests pass. Let me know if there's any more I need to fix/tweak before this can be merged.

voodoos commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -1,9 +1,13 @@
+  | [] | [ _ ] | [ _; _ ] -> file
+  | name :: exts ->
+    let ext = Option.value_exn (List.last exts) in
+    Path.Build.relative dir (sprintf "%s.%s" name ext)
```
I still don't understand why you need to change this function and it's usage at line L253/R273.
Could you document its behavior and why it's needed ?

andreypopp replied:
This is to accommodate [this](https://github.com/andreypopp/dune/blob/10faea361ac82756ebd63a52d8ef66e45270923e/test/blackbox-tests/test-cases/merlin/server.t/run.t#L17-L20) test case.

Before this change we were tracking only config per module name (so extensions were stripped) but now, as we introduce readers which are per extension, we now track config per source file (with extensions).

We still want to treat `module_name.xx.xxx.ext` as `module_name.ext` though as the linked test case specifies, thus the function `strip_pp_extension` normalises a filename by stripping extensions added by a preprocessor.

voodoos replied:
I think I finally understand. I initially though you had the same constraints as preprocessing specification which are per-module (that is, the same for the `ml` and the `mli` file of a module).

But readers can be different for the `ml` file and the `mli` file of a same module ?

andreypopp replied:
> But readers can be different for the ml file and the mli file of a same module ?

You mean for implementation and interface? Yes, I've raise the same question [here](https://github.com/ocaml/dune/pull/8567#issuecomment-1702857875).

But note though that for a single module an implementation might belong to a dialect while the interface could stay pure `.ml` — e.g. `.mlx` and `.mli`. So we still need to track config per file with extension even if we change `(merlin_reader)` to be same both for implementations and interfaces within a dialect.

voodoos replied:
I am not very familiar with Merlin's handling of alternative readers, ~but it looks like the suffix map would allow to use a different reader for the interface and the implementation while keeping a unique configuration per module.~ In fact it's only available as a flag, not a configuration option. So having a configuration for each source file sounds like the easiest way to do it. 👍

voodoos commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -247,35 +272,43 @@ module Processed = struct
-        Path.Build.Map.find per_module_config file_without_ext
+        match Path.Build.Map.find per_module_config file with
+        | Some _ as s -> s
+        | None -> Path.Build.Map.find per_module_config (strip_pp_extensions file)
```
Why do you need this fallbacking mechanism ?

andreypopp replied:
Modules with pp extensions `module_name.xxx.xx.ext` can be also valid modules residing in source directories (e.g. extra extension were not added by a preprocessor but present in source files). In this case it makes sense to try to resolve using non normalised filename first, I think.

voodoos replied:
Do you have a test illustrating that ? In any case, I would like to take some time next week to pull your changes and understand better the relationship between source file / module / merlin config. It's a set of conventions that would benefit a lot from being correctly defined and documented.

The PR looks good otherwise 👍

andreypopp replied:
No, don't have a test case...

1. I'm actually wondering if we should remove any normalisation now that we have per filename config (rather per module) and such config is built by looking at source modules (so we see "real" source files).

1. Another thing to check (add test) is we still support promoted files.

For reference the behaviour of stripping the extension was added in c36df0b73ffa98f893788b25ce68139232299e41

voodoos replied:
> Modules with pp extensions module_name.xxx.xx.ext can be also valid modules residing in source directories (e.g. extra extension were not added by a preprocessor but present in source files). In this case it makes sense to try to resolve using non normalised filename first, I think.

Added a test for that in my commit (at the end of the file), and it doesn't work right now.

> Another thing to check (add test) is we still support promoted files.
That's a good idea, can you do it ? (you can reuse the `merlin_config.sh` script in my new test to easily query the configuration.

rgrinberg commented:
@voodoos are you happy with this PR being merged to be available in 3.11? Or do you want to take another look and have it for 3.12?

voodoos commented:
> @voodoos are you happy with this PR being merged to be available in 3.11? Or do you want to take another look and have it for 3.12?

It's probably all right but I would like to do a last pass and add a few test cases.
What is the schedule for 3.11 ? (I can do that tomorrow)

rgrinberg commented:
I think @emillon wants it out by end of week.

voodoos committed:
test: add a test focusing on merlin's configuration various granularity levels

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>
Signed-off-by: Andrey Popp <8mayday@gmail.com>
Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

voodoos reviewed (dismissed):
I pushed a commit with a new test that focuses on Merlin's config's levels of granularity.

Most standard use cases looks good. There is one change of behavior compared to before: source files with a custom extension that changed after preprocessing are not handled anymore. For example:

Given a file, and a preprocessing action:
`modname.something.cml --pp--> modname.ml` 

One could expect to get the configuration for the module `modname` when running merlin on the `modname.something.cml`.  

Before that PR the granularity was at the module level and a simple heuristic was used to match a source-file with a module: split the filename by `.`, the first segment must be the module name.

Now the granularity is finer, and there is no way, given the source file `modname.something.cml` to guess that the appropriate config is stored under the key `modname.ml`. The new heuristic is: split the filename by `.`, the first segment must be the module name, the last segment must be the extension.

1. I agree that the correct thing to do, as you suggested, @andreypopp, is to use real source-file names as keys:
> I'm actually wondering if we should remove any normalization now that we have per filename config (rather per module) and such config is built by looking at source modules (so we see "real" source files).

2. and adding test for promoted source files sounds important too :-) 

3. Additionally, the currently provided suffixes for melange are `mlx mlx`. Isn't there a different suffix for signatures ? If `mli` are to be used, then I think the list should be `mlx mli` for Merlin jump to declaration to work as expected.

dmmulroy commented:
Are there any updates on this merging? I would love to start seriously utilizing, advocating, and teaching mlx

andreypopp committed:
test: promote merlin tests

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
test: add merlin/dialect.t tests

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
test: tweak the prev commit (to be squashed)

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
refactor: best effort to track orig module sources

This is going to be useful for more granular merlin configs.

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
refactor: per file merlin configs, communicate merlin reader

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
@voodoos I apologise for a prolonged pause, but I've just returned to this PR

I've made the tests cases you've provided pass (thank you for the test cases!) but... it's not that simple.

As I understand it's not possible now to track in dune properly sources for the preprocessed files — actions can be be of any complexity and I imagine a final .ml can have multiple deps/nputs. Not clear which one to treat as a source.

So instead I suggest the following approach:

- we track original source files in `Module.t`
- we construct merlin config by source file and not by module
- we do a look up by the proper filename (with extension)
- ... but we still have a fallback to lookup by `file_without_ext.ml` key — that's exactly to handle preprocessed files with fancy extensions

I don't think this is perfect but it seems to work... 

The ideal solution would be to build a database which would be able to do a lookup by line directives, I think. Similar to how `Copy_line_directive` work but for arbitrary modules, not just those which were `copy#`-ed.

andreypopp committed:
wip: Module.File.t not to encode orig_path

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
in ae596bff74803f9db227607a0619b288f8e09856 I've removed storing `Module.File.orig_path` in `dune-package` but I actually not sure what are the consequences of this change

andreypopp committed:
merlin: lookup fallback with extensionless filename

Relying on `.ml` extension present is not something we can do, instead
store same config with and without extension and on lookup do a fallback
with no extension.

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
in [339d212](https://github.com/ocaml/dune/pull/8567/commits/339d21247cae01ca4630cb64ae3ce9c445bb7041) I've changed the fallback to work like this:
- store same config under original filename and filename w/o any extension as keys
- do a fallback lookup by filename w/o extension

I think this is more robust than doing a fallback `.ml` as extension (not sure it'll always be present)

anmonteiro commented on src/dune_rules/dialect.ml:
```diff
@@ -33,17 +34,19 @@ module File_kind = struct
                 "print_ast"
                 Action.encode
                 (Option.map ~f:(fun (_, x) -> x) print_ast)
+            ; field_o "merlin_reader" (list string) (Option.map ~f:snd merlin_reader)
```
I think we should expect to find:

- 1 dialect per extension
- 1 reader per dialect

what am I missing in this invariant?

andreypopp replied:
the list represents a program name and a list of arguments, so it's a single reader being defined

anmonteiro commented on src/dune_rules/merlin/merlin.ml:
```diff
@@ -563,22 +586,25 @@ module Unprocessed = struct
-          (src, config) :: acc))
-      |> Path.Build.Map.of_list_exn
+          (* we add the config with and without the extension, the latter is
+             needed for a fallback in get *)
```
```suggestion
             needed for a fallback in this file's [get] function. *)
```

anmonteiro commented on src/dune_rules/dialect.ml:
```diff
@@ -271,43 +288,60 @@ module DB = struct
-        then s
-        else { Ml_kind.Dict.impl; intf } :: s)
-      |> List.sort ~compare:(Ml_kind.Dict.compare (Option.compare String.compare))
+  let compute_for_merlin by_name =
+    let extensions, readers =
+      String.Map.fold
+        by_name
+        ~init:([], String.Map.empty)
+        ~f:(fun d (extensions, readers) ->
+          let handle_ml_kind kind readers =
+            let ext = extension d kind in
+            if ext = extension ocaml kind
+            then (* this is standard dialect, exclude *) None, readers
+            else (
+              let reader = merlin_reader d kind in
+              match ext, reader with
+              | Some ext, Some reader -> Some ext, String.Map.add_exn readers ext reader
+              | _ ->
+                if preprocess d kind <> None
+                then (* we have preprocessor defined *) None, readers
+                else ext, readers)
+          in
+          let impl, readers = handle_ml_kind Ml_kind.Impl readers in
+          let intf, readers = handle_ml_kind Ml_kind.Intf readers in
+          let extensions =
+            match impl, intf with
+            | None, None -> extensions
+            | _ -> { Ml_kind.Dict.impl; intf } :: extensions
+          in
+          extensions, readers)
```
we can extract `handle_ml_kind` to the outside scope.
```suggestion
let compute_for_merlin =
  let handle_ml_kind ~dialect kind readers =
    let ext = extension dialect kind in
    if ext = extension ocaml kind
    then (* this is standard dialect, exclude *) None, readers
    else (
      match ext, merlin_reader dialect kind with
      | Some ext, Some reader -> Some ext, String.Map.add_exn readers ext reader
      | _ ->
        if preprocess dialect kind <> None
        then (* we have preprocessor defined *) None, readers
        else ext, readers)
  in
  fun by_name ->
    let extensions, readers =
      String.Map.fold
        by_name
        ~init:([], String.Map.empty)
        ~f:(fun dialect (extensions, readers) ->
          let impl, readers = handle_ml_kind ~dialect Ml_kind.Impl readers in
          let intf, readers = handle_ml_kind ~dialect Ml_kind.Intf readers in
          let extensions =
            match impl, intf with
            | None, None -> extensions
            | _ -> { Ml_kind.Dict.impl; intf } :: extensions
          in
          extensions, readers)
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/dialect.t/dune-project:
```diff
@@ -0,0 +1,10 @@
+(lang dune 3.11)
```
```suggestion
(lang dune 3.16)
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+  $ ocamlc_where="$(ocamlc -where)"
+  $ export BUILD_PATH_PREFIX_MAP="/OCAMLC_WHERE=$ocamlc_where:$BUILD_PATH_PREFIX_MAP"
+
+# A utility to query merlin configuration for a file:
```
```suggestion
A utility to query merlin configuration for a file:
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+
+  $ chmod a+x merlin_conf.sh
+
+# Project sources
```
```suggestion
Project sources
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+# Project sources
+
+  $ cat >dune-project <<EOF
+  > (lang dune 3.11)
```
```suggestion
  > (lang dune 3.16)
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+  44
+  45
+
+# We now query Merlin configuration for the various source files:
```
```suggestion
We now query Merlin configuration for the various source files:
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+Note that suffixes provided by dialect should always be told to Merlin for
+locate to work correctly. 
+
+## Preprocessing:
```
```suggestion
Preprocessing:
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+Dialects are specified by extensions so are specific to a file. This means a
+different reader might be used for the signature and the implementation.
+
+Note that suffixes provided by dialect should always be told to Merlin for
+locate to work correctly. 
```
```suggestion
Note that Merlin should always be told about dialect-provided suffixes, to make `MerlinLocate` work correctly.
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+
+  $ ./merlin_conf.sh pped.mli | diff pped.out -
+
+## Melange:
```
```suggestion
Melange:
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+  \ No newline at end of file
+  [1]
+
+## Unconventionnal file names:
```
```suggestion
Unconventional file names:
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+
+## Unconventionnal file names:
+
+Users might have preprocessing steps that start with a non-conventionnal
```
```suggestion
Users might have preprocessing steps that start with a non-conventional
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+
+  $ ./merlin_conf.sh cppomod.ml | diff cppomod.out -
+
+Note that this means unreleated files might be given the same configuration:
```
```suggestion
Note that this means unrelated files might be given the same configuration:
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+
+  $ ./merlin_conf.sh cppomod.tralala.ml | diff cppomod.out -
+
+And with unconventionnal extension: 
```
```suggestion
And with unconventional extension: 
```

anmonteiro commented on test/blackbox-tests/test-cases/merlin/granularity.t:
```diff
@@ -0,0 +1,178 @@
+  (?:FLG(?:-open?:Dune__exe)
+  (?:FLG(?:-pp?:$TESTCASE_ROOT/_build/default/pp.sh)
+  (?:FLG(?:-w?:@1..3@5..28@30..39@43@46..47@49..57@61..62@67@69-?:-strict-sequence?:-strict-formats?:-short-paths?:-keep-locs?:-no-strict-formats?:-g)
+  (?:SUFFIX?:.mlx .mlx))
```
what does `(?:SUFFIX?:.mlx .mlx))` mean here? Is `pped.ml` getting read by merlin as if it belonged to the mlx dialect?

andreypopp replied:
merlin uses `SUFFIX` config to determine if this is an implementation or an interface document

now that `.mlx .mlx` is not completely correct as mlx dialect doesn't really have an interface counterpart to `.mlx` and it is supposed to use `.mli` instead

@voodoos suggested [here](https://github.com/ocaml/dune/pull/8567#pullrequestreview-1624199713) to use `.mlx .mli` which I want to do but in a separate PR (the code is [here](https://github.com/andreypopp/dune/blob/a5fc103dda82cffff1985dcfff8955ff035b12be/src/dune_rules/merlin/merlin.ml#L144-L150)) as it seems another kind of change.

anmonteiro commented:
@andreypopp took a first look at this change. The approach seems fine, but I have a few questions throughout. I think the main thing I don't understand is why `merlin_reader` is parsed into a list of strings. Is there a case where we would pass multiple readers in that field?

andreypopp committed:
Apply suggestions from code review

Co-authored-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>
Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Extract `handle_ml_kind` to the outside scope.

Co-authored-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>
Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Use Filename.Extension.t for file extensions

Co-authored-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>
Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
tests: update merlin/dialect tests

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
The remaining test failures look like:
```
-  (?:FLG(?:-w?:@1..3@5..28@30..39@43@46..47@49..57@61..62@67@69-?:-strict-sequence?:-strict-formats?:-short-paths?:-keep-locs?:-no-strict-formats?:-g)
+  (?:FLG(?:-w?:@1..3@5..28@31..39@43@46..47@49..57@61..62@67@69-?:-strict-sequence?:-strict-formats?:-short-paths?:-keep-locs?:-no-strict-formats?:-g)
```

note `-w ...28@30...` vs ``-w ...28@31...` and it's changed only since I've bumped dune lang to 3.16... not sure why it has changed...

rgrinberg commented on src/dune_rules/module.ml:
```diff
@@ -3,6 +3,9 @@ open Import
 module File = struct
   type t =
     { path : Path.t
+    ; orig_path : Path.t
```
`original_path`

rgrinberg commented on src/dune_rules/module.mli:
```diff
@@ -88,6 +88,7 @@ module Obj_map : sig
 end
 
 val sources : t -> Path.t list
+val orig_sources : t -> Path.t list
```
`original_sources`. Or better yet, `sources_without_pp`. The originality of a source isn't descriptive without context

andreypopp committed:
More descriptive naming in Module

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
Thanks for the review, @rgrinberg! Tweaked the naming and also rebased on top of `main` to solve conflicts.

andreypopp committed:
promote tests

due to dune 3.16

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
> note -w ...28@30... vs -w ...28@31... and it's changed only since I've bumped dune lang to 3.16... not sure why it has changed...

Javi pointed me to: https://github.com/ocaml/dune/blob/bc57ae438eca9f22f1733dbd58df83e22ee7eaaf/src/dune_rules/ocaml_flags.ml#L34-L39 so I've promoted the tests for merlin granularity.

anmonteiro reviewed (approved):
This looks good to me now. 

@andreypopp feel free to tag me on the next PR about `mlx/mlx` (merlin suffix) too.

andreypopp committed:
merlin: test merlin config for promoted modules

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
Thanks @anmonteiro!

I've added another test case for promoted modules.

@rgrinberg @voodoos please let me know if anything else should be done with this PR or we can merge it?

andreypopp committed:
doc: describe changes

Signed-off-by: Andrey Popp <8mayday@gmail.com>

rgrinberg commented:
No issues from me - though I haven't looked at it closely. @anmonteiro feel free to merge if you're satisfied and if the CI failures aren't related.

anmonteiro committed:
Merge branch 'main' into main

anmonteiro commented:
Let's do this. Thanks for your continued work on this, @andreypopp
